src/digest/base.cr

```diff
@@ -1,92 +1,153 @@
     end
   end
 
-  # Yields a context object with an `#update(data : String | Bytes)`
-  # method available. Returns the resulting digest afterwards.
-  #
-  # ```
-  # require "digest/md5"
-  #
-  # digest = Digest::MD5.digest do |ctx|
-  #   ctx.update "f"
-  #   ctx.update "oo"
-  # end
-  # digest.to_slice.hexstring # => "acbd18db4cc2f85cedef654fccc4a4d8"
-  # ```
-  def self.digest
-    context = new
-    yield context
-    context.final
-    context.result
+  @finished = false
+
+  def update(data) : self
+    update data.to_slice
   end
 
-  # Returns the hexadecimal representation of the hash of *data*.
-  #
-  # ```
-  # require "digest/md5"
-  #
-  # Digest::MD5.hexdigest("foo") # => "acbd18db4cc2f85cedef654fccc4a4d8"
-  # ```
-  def self.hexdigest(data) : String
-    hexdigest &.update(data)
+  def update(data : Bytes) : self
+    check_finished
+    update_impl data
+    self
   end
 
-  # Yields a context object with an `#update(data : String | Bytes)`
-  # method available. Returns the resulting digest in hexadecimal representation
-  # afterwards.
+  # Returns the final digest output.
   #
-  # ```
-  # require "digest/md5"
+  # This method can only be called once and raises `FinalizedError` on subsequent calls.
   #
-  # Digest::MD5.hexdigest("foo") # => "acbd18db4cc2f85cedef654fccc4a4d8"
-  # Digest::MD5.hexdigest do |ctx|
-  #   ctx.update "f"
-  #   ctx.update "oo"
-  # end
-  # # => "acbd18db4cc2f85cedef654fccc4a4d8"
-  # ```
-  def self.hexdigest : String
-    hashsum = digest do |ctx|
-      yield ctx
-    end
+  # NOTE: `.dup.final` call may be used to get an intermediate hash value.
+  def final : Bytes
+    dst = Bytes.new digest_size
+    final dst
+  end
 
-    hashsum.to_slice.hexstring
+  def final(dst : Bytes) : Bytes
+    check_finished
+    @finished = true
+    final_impl dst
+    dst
   end
 
-  # Returns the base64-encoded hash of *data*.
-  #
-  # ```
-  # require "digest/sha1"
-  #
-  # Digest::SHA1.base64digest("foo") # => "C+7Hteo/D9vJXQ3UfzxbwnXaijM="
-  # ```
-  def self.base64digest(data) : String
-    base64digest &.update(data)
+  def reset : self
+    reset_impl
+    @finished = false
+    self
   end
 
-  # Returns the base64-encoded hash of *data*.
-  #
-  # ```
-  # require "digest/sha1"
-  #
-  # Digest::SHA1.base64digest do |ctx|
-  #   ctx.update "f"
-  #   ctx.update "oo"
-  # end
-  # # => "C+7Hteo/D9vJXQ3UfzxbwnXaijM="
-  # ```
-  def self.base64digest : String
-    hashsum = digest do |ctx|
-      yield ctx
-    end
+  # Dups and finishes the digest.
+  @[Deprecated("Use `final` instead.")]
+  def digest : Bytes
+    dup.final
+  end
 
-    Base64.strict_encode(hashsum)
+  # Returns a hexadecimal-encoded digest.
+  @[Deprecated("Use `final.hexstring` instead.")]
+  def hexdigest : String
+    digest.hexstring
   end
+
+  private def check_finished : Nil
+    raise FinalizedError.new("finish already called") if @finished
+  end
+
+  abstract def update_impl(data : Bytes) : Nil
```

Comment:
I'd make the _impl methods private. They're not part of the public API and are only relevant for implementing classes. That would mean they don't show up in the API docs, but I'd rather add implementation instructions in the class documentation.

And `reset_impl` needs an abstract def.

Reply:
I'm just copying what's in the crystal repo.

Other inherited classes don't document the abstract methods in the class documentation but keep it with the abstract method.

```
# Flushes the wrapped `IO`.
abstract def unbuffered_flush
```

They're also not `private`.  Maybe so they do show up in the base class documentation avoiding duplication of documenting the definitions, types, and return values?

Reply:
> They're also not `private`. Maybe so they do show up in the base class documentation avoiding duplication of documenting the definitions, types, and return values?

I believe it's because you can't use visibility modifiers within the `abstract def` declaration - they have to be on the implementation side only.

Reply:
`abstract def unbuffered_read` shows in `IO::Buffered` documentatation.

`src/http/server/response.cr:198:      private def unbuffered_read(slice : Bytes)` doesn't.

Is that not the correct way to do it?  It seems logical to me.  

If the `abstract def`'s are private I have to duplicate the definitions for every `_impl` method in the class documentation.  Which may get out of sync over time.

If that's not the correct way why aren't all the other method definitions changed?

Reply:
No, you can define `private abstract` methods.

The public methods are relevant for all users. But implementation methods are only relevant when implementing the interface, they're not supposed to be called on their own. So they should not be listed in the public interface. 

The purpose can differ depending on the API in question. But some implementations of `IO::Buffered` actually declare the methods implementations as `private`. They're simply not relevant for using it. Maybe we should change that for `IO::Buffered` abstract methods as well. But that's a separate issue.

Reply:
I think that if an abstract method is supposed to be implemented by subclasses, and subclasses can be defined outside of the std, then they should be public, or maybe protected. But if they are protected or private they won't show up in docs, so maybe public is fine? In any case I don't think anyone will invoke those methods.

Or put another way: maybe it doesn't matter too much?

Reply:
I think it matters when you have a bunch of duplicate methods in the API docs where only halve of them are meant to be called from user code.